### ISSUE TITLE
TitleScreen module changing the bank for the Kirby sprite assembly (Debug menu)

### ISSUE THREAD

**Issue by ShadowOne333** [mrtenda/CoilSnake#166]
The Title Screen module changes one load from accumulator which specifies a certain data bank for the character layout (CHARS_LAYOUT_BANK = 0xA0FE), which is also required for the Kirby sprite assembly to properly output the Kirby sprite correctly into the Boot-Up Debug Menu.

The module at hand which is related to the problem:
https://github.com/mrtenda/CoilSnake/blob/master/coilsnake/modules/eb/TitleScreenModule.py

Lines of interest:
**55, 182-191, 276-315**

CoilSnake modifies two bytes in address 0x00A0FE, in the original EarthBound it's [A5 8E], or LDA $8E, while in a compiled hack, these bytes could change, for example in MaternalBound these two bytes were changed to [A9 DE], or LDA #$DE.
These two bytes tell the game in what data bank to look for the Character Layouts, but the Kirby sprite assembly also depends on this, as right now with any compiled ROM it is pulling the sprite assembly from the wrong data bank.
Kirby’s sprite assembly data is located at $EFF5BE (rom offset $2FF5BE) for reference.

As a side effect to this, the animation for the letters closing in for the Title screen do not work if the bank changes through Hex editing, so the way to make both of them work is to somehow make sure that both of these datas, the Kirby sprite assembly and the character layout bank, have the same data bank with one another.

### **Suggested Solution:**

A modification to the TitleScreen module from CoilSnake’s source to also decompile/compile the Kirby sprite, so that both the character layouts and the Kirby sprite assembly use the same data bank. 
Here are some addresses that are related to the Kirby sprite.

**Offsets of interest:**
(All PC addresses are for a headerless ROM, add 0x200 if using a headered ROM)

- 2FEFB7-2FF0D6 (000120) = Programmers' Debug Mode Selection Kirby Cursor Graphics (EFEFB7-EFF0D6 in SNES address)

- 2FF7BE (EFF5BE in SNES address) - Kirby’s sprite assembly data

- 2FE75E (EFE55E in SNES address) - Kirby graphics pointer
[ A9 B7 EF 85 0E A9 EF 00 85 10 ]

**Comment by ShadowOne333** [comment]
_(Following up on the issue)_

I have managed to make a somewhat consistent fix for the sprite in MaternalBound.
However, the only drawback is that each hack will have a different bank assigned on compilation, but seems that once the hack is compiled once, the bank tends to always be the same. I have compiled MaternalBound 5+ times with this fix, and the bank hasn't changed in each one, though I still have yet to see if making more changes to the hack would make it change the bank again.

Now unto the fix itself:
```
/* KIRBY SPRITE ASSEMBLY */
// This part focuses on fixing the Kirby sprite cursor that appears on the Boot-Up Debug Menu

// Assembly 3-byte pointer
ROM[0xC2FFB8] = "[D0 FF DE]"

// Assembly 2-byte pointer
ROM[0xDEFFD0] = "[D2 FF]"

// Assembly  - Data bank determined by LDA opcode at 0xA0FE
ROM[0xDEFFD2] = "[F4 00 30 F4 00 F4 01 30 FC 00 F4 02 30 04 00 FC 03 30 F4 00 FC 04 30 FC 00 FC 05 30 04 00 04 06 30 F4 00 04 07 30 FC 00 04 08 30 04 80]"
```

This is all the code that I used for MaternalBound to fix the sprite.
To explain each part:

- `ROM[0xC2FFB8] = "[D0 FF DE]"`
This is the 3 byte pointer that tells the game where to look for the Kirby sprite assembly in the ROM (in SNES address).
The part of importance here is the last byte, **DE** in my case, since this HAS to match the data bank that is found at _0x00A0FE_ once the ROM is compiled. For MaternalBound, this seems to always be _LDA #$DE_, so this has to match with that byte of the pointer. The rest [D0 FF] is a random address that can be changed inside the DE bank specified before. I simply used that address because it's at the very end of the bank, to make it less likely to collide with other data (like the Title Screen data).
**NOTE:** The data bank ($DE for MaternalBound) changes with each hack! This one has to match with whatever data bank is assigned at 0x00A0FE, which gets determined either at [Line 186 in the TitleScreenModule.py](https://github.com/mrtenda/CoilSnake/blob/master/coilsnake/modules/eb/TitleScreenModule.py#L186), or somewhere [in between Line 276 and Line 315](https://github.com/mrtenda/CoilSnake/blob/master/coilsnake/modules/eb/TitleScreenModule.py#L276).

- `ROM[0xDEFFD0] = "[D2 FF]"`
This is a 2 byte pointer. The 3 byte pointer goes to the start of this data, which simply redirects to the start of the sprite assembly two bytes after this pointer (odd, I know).

- `ROM[0xDEFFD2] = "[F4 00 30 F4 00 F4 01 30 FC 00 F4 02 30 04 00 FC 03 30 F4 00 FC 04 30 FC 00 FC 05 30 04 00 04 06 30 F4 00 04 07 30 FC 00 04 08 30 04 80]"`
This is the sprite assembly itself.
This can be treated as simple hex data, or converted to ASM 65816. However, the assembly code for this is... interesting to say the least. I would recommend that if this is going to be implemented into the title screen module, to use it as raw hex data, instead of assembly, to avoid possible issues.


This is all that is needed to successfully restore the Kirby sprite. The sprite itself, as well as its pointer, are not needed in this case, so that can be omitted.
